Switch ocean internally to use ocean.meta

mihails-strasuns-sociomantic opened the pull request:
First step towards eventual deprecation of ocean.core.Traits

mihails-strasuns-sociomantic commented:
(and last big change for v4.0.0)

david-eckardt-sociomantic commented on src/ocean/core/TypeConvert.d:
```diff
@@ -192,7 +191,7 @@ template arrayOf (T)
     {
         static assert (U.length > 0);
-        static assert (!hasIndirections!(U));
+        static assert (!hasIndirections!(U[0]));
```
Is this a separate bugfix?

david-eckardt-sociomantic commented on src/ocean/text/convert/Formatter.d:
```diff
@@ -532,8 +536,10 @@ private void handle (T) (T v, FormatInfo f, FormatterSink sf, ElemSink se)
 
     // Arrays (dynamic and static)
-    else static if (is (T A : A[]))
+    else static if (isBasicArrayType!(T))
```
Does
```d
 static if (is(ElementTypeOf!(T) A))
```
work? ;)

mihails-strasuns-sociomantic replied:
Yep (great idea btw!), but it is not very suitable here as `ElementTypeOf` also works on associative arrays.

david-eckardt-sociomantic commented on src/ocean/util/DeepReset.d:
```diff
@@ -186,18 +190,18 @@ public void StructDeepReset ( T ) ( ref T dst )
                 DynamicArrayDeepReset(dst.tupleof[i]);
             }
-            else // static array
+            else
```
Maybe refactor this to
```d
else static if ( isArrayType!(typeof(member)) == ArrayKind.Dynamic )
    DynamicArrayDeepReset(dst.tupleof[i]);
else static if ( isArrayType!(typeof(member)) == ArrayKind.Static )
    StaticArrayDeepReset(dst.tupleof[i]);
```

david-eckardt-sociomantic commented on src/ocean/util/config/ConfigFiller.d:
```diff
@@ -671,38 +669,31 @@ struct SetInfo ( T )
+        return true;
     else static if ( isIntegerType!(T) || isRealType!(T) )
+        return true;
+    else static if ( isArrayType!(T) )
```
If `static if (is(ElementTypeOf!(T) U))` works then you don't need to add the `IsSupportedImpl` function.

david-eckardt-sociomantic commented on src/ocean/util/config/ConfigFiller.d:
```diff
@@ -1047,7 +1038,7 @@ private void readFieldsImpl ( T, Source )
         if ( config.exists(group, key) )
         {
-            static if ( is(Type U : U[]) && !isStringType!(Type))
+            static if (is(Type U : U[]) && !isUTF8StringType!(Type))
```
I'm surprised you keep using `is(Type U : U[])` here.

mihails-strasuns-sociomantic replied:
My primary goal was right now is to stop using `core.Traits`, not improve generic facilities in general (that is a bit more complicated task), so I haven't been paying much attention to cases like this.

david-eckardt-sociomantic replied:
Fair enough.

mihails-strasuns-sociomantic committed:
Use `ocean.meta` in `arrayOf` and adjust impl

Old `hasIndirection` was accepting arbitrary amount of template
arguments and checking them all. New one expects only one argument and
in `arrayOf` case checking only `U[0]` is generally sufficient.

mihails-strasuns-sociomantic committed:
Copy `TemplateInstanceArgs` to `ocean.meta`

mihails-strasuns-sociomantic committed:
Switch ocean internally to use ocean.meta

First step towards eventual deprecation of ocean.core.Traits

mihails-strasuns-sociomantic commented:
Rebased, updated per @david-eckardt-sociomantic comments, fixed D2 build.

david-eckardt-sociomantic reviewed (approved):
Ick kann nich meckern.
